File: sdk/java/driversdk/src/main/java/com/baidu/bce/sdk/utils/ParseValue.java
package com.baidu.bce.sdk.utils;

import com.baidu.bce.sdk.context.models.device.ArrayType;
import com.baidu.bce.sdk.context.models.device.EnumType;
import com.baidu.bce.sdk.context.models.device.EnumValue;
import com.baidu.bce.sdk.context.models.device.ObjectType;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

public class ParseValue {
    public static final String TYPE_INT = "int";
    public static final String TYPE_INT16 = "int16";
    public static final String TYPE_INT32 = "int32";
    public static final String TYPE_INT64 = "int64";
    public static final String TYPE_FLOAT32 = "float32";
    public static final String TYPE_FLOAT64 = "float64";
    public static final String TYPE_BOOL = "bool";
    public static final String TYPE_STRING = "string";
    public static final String TYPE_TIME = "time";
    public static final String TYPE_DATE = "date";
    public static final String TYPE_ARRAY = "array";
    public static final String TYPE_ENUM = "enum";
    public static final String TYPE_OBJECT = "object";

    private static String[] dateLayout = {
            "yyyy-MM-dd", "yyyy.MM.dd", "yyyy/MM/dd"
    };

    private static String[] timeLayout = {
            "HH:mm:ss", "HH-mm-ss", "HH.mm.ss"
    };

    public static Object parse(String typ, Object value, Object args) throws ParseException {
        try {
            return switch (typ) {
                case TYPE_INT, TYPE_INT32 -> Integer.parseInt(value.toString());
                case TYPE_INT16 -> Short.parseShort(value.toString());
                case TYPE_INT64 -> Long.parseLong(value.toString());
                case TYPE_FLOAT32 -> Float.parseFloat(value.toString());
                case TYPE_FLOAT64 -> Double.parseDouble(value.toString());
                case TYPE_BOOL -> Boolean.parseBoolean(value.toString());
                case TYPE_STRING -> value.toString();
                case TYPE_DATE -> parseDate(value.toString(), args.toString(), dateLayout);
                case TYPE_TIME -> parseDate(value.toString(), args.toString(), timeLayout);
                case TYPE_ARRAY -> parseArray(value, args);
                case TYPE_ENUM -> parseEnum(value, args);
                case TYPE_OBJECT -> parseObject(value, args);
                default -> null;
            };
        } catch (Exception e) {
            throw new ParseException(e.getMessage(), 0);
        }
    }

    public static Map<String, Object> parseObject(Object obj, Object args) throws ParseException {
        Map<String, Object> res = null;
        if (args instanceof Map && obj instanceof Map) {
            res = new HashMap<>();

            for (Map.Entry<?, ?> entry : ((Map<?, ?>) args).entrySet()) {
                if (entry.getKey() instanceof String && entry.getValue() instanceof ObjectType) {
                    Object originVal = ((Map<?, ?>) obj).get(entry.getKey());
                    res.put(((String) entry.getKey()), parse(((ObjectType) entry.getValue()).getType(),
                            originVal, ((ObjectType) entry.getValue()).getFormat()));
                }
            }
        }
        return res;
    }

    public static String parseEnum(Object obj, Object args) throws ParseException {
        if (args instanceof EnumType) {
            for (EnumValue item : ((EnumType) args).getValues()) {
                Object val = parse(((EnumType) args).getType(), item.getValue(), null);
                if (val.equals(obj)) {
                    return item.getName();
                }
            }
        }
        return "";
    }

    public static Object[] parseArray(Object obj, Object args) throws ParseException {
        Object[] res = null;
        if (args instanceof ArrayType) {
            if (obj instanceof Object[]) {
                if (((Object[]) obj).length <= ((ArrayType) args).getMax()
                        && ((Object[]) obj).length >= ((ArrayType) args).getMin()) {
                    res = new Object[((Object[]) obj).length];
                    for (int i = 0; i < ((Object[]) obj).length; i++) {
                        Object t = parse(((ArrayType) args).getType(),
                                ((Object[]) obj)[i], ((ArrayType) args).getFormat());
                        res[i] = t;
                    }
                }
            }
        }
        return res;
    }

    public static String parseDate(String val, String format, String[] layouts) {
        SimpleDateFormat sdf;
        Date date = null;
        for (String layout : layouts) {
            sdf = new SimpleDateFormat(layout);
            try {
                date = sdf.parse(val);
            } catch (ParseException e) {
                continue;
            }
            break;
        }
        if (date != null) {
            sdf = new SimpleDateFormat(format);
            return sdf.format(date);
        }
        return "";
    }

}

File: sdk/csharp/driversdk/DriverSDK/Utils/ParseValue.cs
using DriverSDK.DmContext.Models;
using System.Globalization;

namespace DriverSDK.Utils
{
    public class ParseValue
    {
        public const string TYPE_INT = "int";
        public const string TYPE_INT16 = "int16";
        public const string TYPE_INT32 = "int32";
        public const string TYPE_INT64 = "int64";
        public const string TYPE_FLOAT32 = "float32";
        public const string TYPE_FLOAT64 = "float64";
        public const string TYPE_BOOL = "bool";
        public const string TYPE_STRING = "string";
        public const string TYPE_TIME = "time";
        public const string TYPE_DATE = "date";
        public const string TYPE_ARRAY = "array";
        public const string TYPE_ENUM = "enum";
        public const string TYPE_OBJECT = "object";

        private static readonly string[] DateLayout = { "yyyy-MM-dd", "yyyy.MM.dd", "yyyy/MM/dd" };
        private static readonly string[] TimeLayout = { "HH:mm:ss", "HH-mm-ss", "HH.mm.ss" };

        public static object? Parse(string typ, object value, object args)
        {
            try
            {
                return typ switch
                {
                    TYPE_INT or TYPE_INT32 => Convert.ToInt32(value),
                    TYPE_INT16 => Convert.ToInt16(value),
                    TYPE_INT64 => Convert.ToInt64(value),
                    TYPE_FLOAT32 => Convert.ChangeType(value, typeof(float)),
                    TYPE_FLOAT64 => Convert.ToDouble(value),
                    TYPE_BOOL => Convert.ToBoolean(value),
                    TYPE_STRING => Convert.ToString(value),
                    TYPE_DATE => ParseDate(Convert.ToString(value), Convert.ToString(args), DateLayout),
                    TYPE_TIME => ParseDate(Convert.ToString(value), Convert.ToString(args), TimeLayout),
                    TYPE_ARRAY => ParseArray(value, args),
                    TYPE_ENUM => ParseEnum(value, args),
                    TYPE_OBJECT => ParseObject(value, args),
                    _ => null,
                };
            }
            catch (Exception e)
            {
                throw new InvalidCastException(e.Message);
            }
        }

        public static Dictionary<string, object>? ParseObject(object obj, object args)
        {
            if (args is IDictionary<string, ObjectType> argsDict && obj is Dictionary<string, object> objDict)
            {
                var res = new Dictionary<string, object>();
                foreach (var entry in argsDict)
                {
                    if (entry.Key is string key && entry.Value is ObjectType objectType && objDict.ContainsKey(key))
                    {
                        var originVal = objDict[key];
                        if (objectType.Type != null && originVal != null && objectType.Format != null)
                        {
                            var parRes = Parse(objectType.Type, originVal, objectType.Format);
                            if (parRes != null)
                            {
                                res[key] = parRes;
                            }
                        }
                    }
                }
                return res;
            }
            return null;
        }

        public static string ParseEnum(object obj, object args)
        {
            if (args is EnumType enumType && enumType.Values != null && enumType.Type != null)
            {
                foreach (var item in enumType.Values)
                {
                    var val = Parse(enumType.Type, item.Value, "");
                    if (val != null && val.Equals(obj))
                    {
                        return item.Name;
                    }
                }
            }
            return "";
        }

        public static object[]? ParseArray(object obj, object args)
        {
            if (args is ArrayType arrayType && obj is object[] objArray)
            {
                if (objArray.Length <= arrayType.Max && objArray.Length >= arrayType.Min)
                {
                    var res = new object[objArray.Length];
                    for (int i = 0; i < objArray.Length; i++)
                    {
                        var t = Parse(arrayType.Type, objArray[i], arrayType.Format);
                        if (t != null)
                        {
                            res[i] = t;
                        }
                    }
                    return res;
                }
            }
            return null;
        }

        public static string ParseDate(string? val, string? format, string[] layouts)
        {
            if (val == null || format == null)
            {
                return "";
            }
            foreach (var layout in layouts)
            {
                if (DateTime.TryParseExact(val, layout, CultureInfo.InvariantCulture, DateTimeStyles.None, out DateTime date))
                {
                    return date.ToString(format);
                }
            }
            return "";
        }
    }
}

